### ISSUE TITLE
Confusing Cancellation in BackgroundService

### ISSUE THREAD

**Issue by DanVicarel** [dotnet/extensions#1245]
### Describe the bug
`BackgroundService` (and really the `IHostedService` interface) have kind of a confusing cancellation model that makes it tricky to test.

### To Reproduce
Derive from `BackgroundService`, e.g.:

```cs
public class DerpService : BackgroundService
{
    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
    {
        while (true) {
            cancellationToken.ThrowIfCancellationRequested();
            await Task.Delay(100, cancellationToken);
        }
    }
}
```

Write some code to test cancellation of the `DerpService`, e.g., with NUnit:

```cs
[Test, Parallelizable]
public async Task SomeTest()
{
    var derp = new DerpService();
    var cts = new CancellationTokenSource(millisecondsDelay: 1000);

    // Returns almost immediately and never throws
    Assert.ThrowsAsync<OperationCanceledException>(() => derp.StartAsync(cts.Token));
}
```

### Expected behavior
After 1 second, the `CancellationTokenSource` in the test method should automatically cancel.  At this point, I would expect the `DerpService` to see the cancellation, throw an `OperationCanceledException`, and thus pass the assertion.  Instead, the call to `StartAsync` just returns after one iteration of delay and never waits for the `CancellationToken`.

### Additional context

Looking at the [source code of `BackgroundService`](https://github.com/aspnet/Extensions/blob/master/src/Hosting/Abstractions/src/BackgroundService.cs), it's clear that this happens because `BackgroundService` defines its _own_ `CancellationTokenSource`, and passes a token from _that_ source to `ExecuteAsync`. I also don't see any unit tests for `BackgroundService`, which is perhaps why this issue hasn't been encountered before.

So I guess awaiting `StartAsync` doesn't really mean awaiting the _service_, it just means awaiting the _start_ of the service, which is kind of confusing.  While I realize that `BackgroundService`, like all implementations of `IHostedService`, can be stopped via `StopAsync`, I don't really see the point of passing a `CancellationToken` to `StartAsync` that gets completely ignored.  For that matter, I don't see why `IHostedService` declares a `StopAsync` method at all.  Any code that runs an `IHostedService` should be able to stop it just by cancelling the token that was passed to `StartAsync`, and implementers could put whatever "stopping" logic they need in a call to `CancellationToken.Register()`.  This would also allow `OperationCanceledException`s raised in the service to bubble up, rather than getting squashed in `StopAsync`.

**Output of `dotnet --info`**:

.NET Core SDK (reflecting any global.json):
 Version:   2.2.103
 Commit:    8edbc2570a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.103\

Host (useful for support):
  Version: 2.2.1
  Commit:  878dd11e62

.NET Core SDKs installed:
  1.0.0-preview2-1-003177 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.502 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.1.504 [C:\Program Files\dotnet\sdk]
  2.1.600-preview-009426 [C:\Program Files\dotnet\sdk]
  2.1.600-preview-009472 [C:\Program Files\dotnet\sdk]
  2.1.600-preview-009497 [C:\Program Files\dotnet\sdk]
  2.1.600 [C:\Program Files\dotnet\sdk]
  2.2.103 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

**Comment by davidfowl** [comment]
> While I realize that BackgroundService, like all implementations of IHostedService, can be stopped via StopAsync, I don't really see the point of passing a CancellationToken to StartAsync that gets completely ignore

It's part of the interface but I can see how you would expect it to do *something*, I'm just not sure what it should do.

> For that matter, I don't see why IHostedService declares a StopAsync method at all. Any code that runs an IHostedService should be able to stop it just by cancelling the token that was passed to StartAsync, and implementers could put whatever "stopping" logic they need in a call to CancellationToken.Register(). This would also allow OperationCanceledExceptions raised in the service to bubble up, rather than getting squashed in StopAsync.

The best analogy here is an API that has an asynchronous way to start an operation and stop an operation, like say connecting to another service. You need to asynchronously Start, which needs to support cancellation and you need to also asynchronously stop (which also supports canceallation for timeout reasons). Those tokens last as long as the Start and Stop operations but have no effect on the thing that was started or stopped. That's an entirely different programming model.

IHostedService.StartAsync - Execute asynchronous startup logic
IHostedService.StopAsync - Execute asynchronous shutdown logic

The background service flips the model and assumes you want to do a long running async operation, essentially, the code that runs between StartAsync (since your background service always starts up synchronously but see other confusion here https://github.com/aspnet/Extensions/issues/813) and StopAsync. The token triggers when the IHostedService.StopAsync is called. That's the contract and it's the reason we made StartAsync and StopAsync virtual so they could be overridden and handled differently (if you wanted to use the token to kick off something asynchronous).

**Comment by davidfowl** [comment]
> The separation between tasks for starting the service, stopping the service, and the service itself makes sense; I just don't see why that's the contract that was decided on for IHostedService. My description of an interface with only a cancellable StartAsync method feels like what most people would want/expect.

Maybe another way to put this is, why does BackgroundService implement IHostedService. Maybe we should have implemented IHostedService explicitly so that you don't get to call StartAsync and StopAsync but only ExecuteAsync that does the long running thing. That would make your test just work and act like you'd expect.

**Comment by DanVicarel** [comment]
>why does BackgroundService implement IHostedService. Maybe we should have implemented IHostedService

I don't understand this.  Did you mean something other than `IHostedService` for the latter?

**Comment by DanVicarel** [comment]
Oh okay nvm, I didn't realize that explicit interface implementation actually hides an implemented method unless you define a variable with the interface type.  That's an interesting idea.  It would make the following code possible:

```cs
public class DerpService : BackgroundService {
        Task IHostedService.StartAsync(CancellationToken cancellationToken) { //... }
        Task IHostedService.StopAsync(CancellationToken cancellationToken) { //... }
}

// ...
DerpService derp = new DerpService();
IHostedService iderp = derp as IHostedService;

derp.StartAsync();    // Compiler error
derp.StopAsync();    // Compiler error
derp.ExecuteAsync();    // Fine

iDerp.StartAsync();    // Fine
iDerp.StopAsync();    // Fine
```

Maybe I'm just complaining about explicit interface implementation now, but that code looks very confusing to me.  Still, I guess it would at least force `BackgroundService` into the "long-running" contract, as you said.

**Comment by anurse** [comment]
It sounds like what we need here is to better clarify a pattern for testing background services. To me, the pattern is more like this:

```csharp
public async Task MyTest()
{
    var service = /* get my background service */
    await service.StartAsync();
    await Task.Delay(/* enough time for the service to run */);
    await service.StopAsync();
}
```

Alternatively, rather than using a timer, you could have your background service signal something to tell your test that it has finished it's work.

We should improve the background service doc to add some testing notes. We should also clarify the token usage in XML docs.

**Comment by DanVicarel** [comment]
@anurse  I like both of those ideas.

**Comment by Zenexer** [comment]
@DanVicarel, there are really two ways in which an IHostedService is likely to be implemented.  BackgroundService covers one of those, but it sounds a bit like you're either expecting the alternative method or are losing sight of the context in which an IHostedService is likely to run.

If you've worked with IHostedServices a bit, you've probably noticed that any exceptions that bubble beyond StartAsync/StopAsync get lost.  This is because an IHostedService is usually running very close to the entrypoint of the application.  Any cancellation tokens it receives directly are likely to be useless because they would only indicate that the application is exiting.  You can pass a CancellationToken if you're doing something a bit less common and have a legitimate reason to cancel the start/stop operations, but chances are it's not going to be useful.

If an IHostedService is wrapping a more traditional service, it's likely going to simply attach a listener when StartAsync is called, then return Task.CompletedTask.  It may additionally detach the listener or dispose the underlying service when StopAsync is called.  The key observation to note here is that StartAsync and StopAsync are both short-lived.  StartAsync will run to completion before the IHostedService has logically stopped.  The Task returned by StartAsync will most likely run synchronously and indicate immediate completion.

If an IHostedService is going to be performing a long-running operation using the more modern async/await paradigm, then StartAsync is expected to be where that long-running operation occurs.  The Task returned by StartAsync will run indefinitely and will only indicate completion if the IHostedService is stopped.  This is where BackgroundService comes in: there needs to be a way to cancel the task from StartAsync when StopAsync or Dispose are called.  BackgroundService's job is to manage a CancellationTokenSource so that it can cancel the token that's passed to ExecuteAsync.  It can't cancel the token that's passed to StartAsync, since that's a CancellationToken, not a CancellationTokenSource, so it has to pass something else to ExecuteAsync.

There are still some issues with exactly how BackgroundService manages its CancellationTokenSource which I'm looking into addressing in another issue.  I'll link to this issue from the new issue.

**Comment by Zenexer** [comment]
Additional note I'd like to make:

The existence of BackgroundService indicates that no assumption is made about the purpose of the task returned by StartAsync: the task may be exclusively responsible for starting the service, or it may be responsible for executing the service.  This in turn means that the cancellationToken passed to StartAsync may cancel the start operation, or it may cancel the whole service, with a caller unable to determine which will happen.

This means IHostedService makes no guarantees about the effect of canceling the token passed to StartAsync.  It's essentially a useless parameter and should always be `default(CancellationToken)`, since a caller has no way of knowing what effect cancellation will have.  This seems like a design flaw, and IHostedService should really be split into two different interfaces.

In theory, BackgroundService could be modified to always return Task.CompletedTask for StartAsync instead of the result of ExecuteAsync.  However, since it has already been widely used as an example, the expectations for IHostedService.StartAsync's return value couldn't change.  To avoid breaking existing expectations, two new interfaces should be introduced to replace IHostedService.

**Comment by haakoo** [comment]
Based on the examples in this repo and in the AspNetCore.Docs repo I've interpreted `IGenericHost.StartAsync` as starting the server, and a completed task means the service is running.

**Comment by Zenexer** [comment]
@haakoo If I'm not mistaken, it's not safe to make any assumption as to what the completion of the Task returned by StartAsync means.  It may mean that the service has started, or it may mean that the service is stopping/has stopped.

By extension, a Task that is running may indicate that the service is starting, or it may indicate that the service is running.

**Comment by MV10** [comment]
> If I'm not mistaken, it's not safe to make any assumption as to what the completion of the Task returned by StartAsync means. It may mean that the service has started, or it may mean that the service is stopping/has stopped.

@Zenexer I suspect that's inaccurate, if you look at the bigger picture of the host lifecycle. I believe `IHostApplicationLifetime` only triggers `ApplicationStarted` (a cancellation token, not an event) after all `StartAsync` calls have been completed, although I haven't tracked it down in the source to be absolutely sure, this is all based on log output.

It looks to me like the correct usage is to `Register` the call to your async "do work" method with the `ApplicationStarted` token, and to only use `StartAsync` as a kind of signal to initialize the service. Otherwise your background work begins before other services have even have a chance to start. If there are dependencies across the hosted services, this is a race condition waiting to happen.

Basing start-of-work on `ApplicationStarted` yields a predictable sequence of events, otherwise you have jobs starting and potentially finishing/failing before app startup is even signaled. I went through this yesterday with a set of hosted services that start and monitor external processes. One failed almost immediately and called for app shutdown before app startup was logged. It took awhile to figure out what was happening, but after switching all the services to use `ApplicationStarted`, every combination I've tried results in a sensible sequence of events.

Another indication (to me) that completion of the `StartAsync` task isn't interpreted as the service running to completion is that if you host just a single service, returning `CompletedTask` won't terminate the host. You still have to explicitly call `StopApplication` (or CTRL+C or whatever).

If that's all correct, then I suspect `BackgroundService` might be incorrect (as you noted earlier, "StartAsync's task completion indicates that the service has stopped").

**Comment by Zenexer** [comment]
It cannot be considered incorrect at this point because it has been widely used as an example, so there is a lot of code out in the wild that behaves similarly. Without breaking backwards compatibility, the only safe option is to make no assumptions about the Task returned by StartAsync: when it completes, it may indicate that the service has finished starting, or it may indicate that the service has run to completion. This ambiguity is unfortunate.

Registering an async method with a CancellationToken is not safe. CancellationToken.Register doesn’t understand Tasks and won’t await the result. Furthermore, if you have long running tasks, as would be expected from a service, there’s a good chance your service won’t yield immediately, which prevents other registered delegates from running. Tasks need to be awaited; BackgroundService doesn’t do this properly, which is a separate bug that can cause memory leaks and lost exceptions. I’m on my phone, so it’s difficult to find, but I have an open issue for that.  If registering with ApplicationStarted appears to resolve a problem with your service, it’s likely that the serial nature of CancellationTokens is just hiding a deeper concurrency issue with your service.

**Comment by MV10** [comment]
It isn't awaited, but my thinking was that it's effectively the same reason async void is valid for event handlers and callbacks: the _caller_ (here, token Register) doesn't care about the outcome of the Task. After all, async void event handlers aren't awaited either.

**Comment by babelshift** [comment]
Is this related to https://github.com/dotnet/extensions/pull/2823?

**Comment by GeeWee** [comment]
I still think this is relevant.

**Comment by anurse** [comment]
I believe this has actually been fixed in 5.0. The token ExecuteAsync receives is linked to the one passed to StartAsync. So ExecuteAsync’s token will be cancelled if EITHER of the below happen:

- StartAsync’s token is cancelled
- StopAsync is called

Can you give that a try and see if it resolves your issue? It should be present in the latest 5.0 preview.

**Comment by Zenexer** [comment]
@anurse Do you know if the fix is likely to have an impact on #2017?

**Comment by joskoanicic** [comment]
> @anurse Do you know if the fix is likely to have an impact on #2017?
@Zenexer 
Looking at the code https://github.com/dotnet/extensions/pull/2823, no. 

Issue you are mentioning is still relevant.

**Comment by anurse** [comment]
Closing this issue as the original suggestion was completed in 5.0. #2017 is not affected, and is still open to track work there.
